hibernate-core/src/test/java/org/hibernate/test/subselectfetch/SubselectFetchCollectionFromBatchTest.java

```diff
@@ -0,0 +1,267 @@
+
+import static org.junit.Assert.assertEquals;
+import static org.junit.Assert.assertFalse;
+import static org.junit.Assert.assertTrue;
+
+/**
+ * @author Stephen Fikes
+ * @author Gail Badner
+ */
+public class SubselectFetchCollectionFromBatchTest  extends BaseCoreFunctionalTestCase {
+
+	@Override
+	public void configure(Configuration cfg) {
+		cfg.setProperty( Environment.GENERATE_STATISTICS, "true");
+	}
+
+	@Test
+	public void testSubselectFetchFromEntityBatch() {
+		Session s = openSession();
+		Transaction t = s.beginTransaction();
+		EmployeeGroup group1 = new EmployeeGroup();
+		Employee employee1 = new Employee("Jane");
+		Employee employee2 = new Employee("Jeff");
+		group1.addEmployee(employee1);
+		group1.addEmployee(employee2);
+		EmployeeGroup group2 = new EmployeeGroup();
+		Employee employee3 = new Employee("Joan");
+		Employee employee4 = new Employee("John");
+		group2.addEmployee(employee3);
+		group2.addEmployee( employee4 );
+
+		s.save( group1 );
+		s.save( group2 );
+		s.flush();
+
+		s.clear();
+
+		sessionFactory().getStatistics().clear();
+
+		EmployeeGroup[] groups = new EmployeeGroup[] {
+				(EmployeeGroup) s.load(EmployeeGroup.class, group1.getId()),
+				(EmployeeGroup) s.load(EmployeeGroup.class, group2.getId())
+		};
+
+		// groups should only contain proxies
+		assertEquals( 0, sessionFactory().getStatistics().getPrepareStatementCount() );
+
+		for (EmployeeGroup group : groups) {
+			assertFalse( Hibernate.isInitialized( group ) );
+		}
+
+		assertEquals( 0, sessionFactory().getStatistics().getPrepareStatementCount() );
+
+
+		for (int i = 0 ; i < groups.length; i++ ) {
+			// Both groups get initialized  and are added to the PersistenceContext when i == 0;
+			// Still need to call Hibernate.initialize( groups[i] ) for i > 0 so that the entity
+			// in the PersistenceContext gets assigned to its respective proxy target (is this a
+			// bug???)
+			Hibernate.initialize( groups[ i ] );
+			assertTrue( Hibernate.isInitialized( groups[i] ) );
+			// the collections should be uninitialized
+			assertFalse( Hibernate.isInitialized( groups[i].getEmployees() ) );
+		}
+
+		// both Group proxies should have been loaded in the same batch;
+		assertEquals( 1, sessionFactory().getStatistics().getPrepareStatementCount() );
+		sessionFactory().getStatistics().clear();
+
+		for (EmployeeGroup group : groups) {
+			assertTrue( Hibernate.isInitialized( group ) );
+			assertFalse( Hibernate.isInitialized( group.getEmployees() ) );
+		}
+
+		assertEquals( 0, sessionFactory().getStatistics().getPrepareStatementCount() );
+
+		// now initialize the collection in the first; collections in both groups
+		// should get initialized
+		Hibernate.initialize( groups[0].getEmployees() );
+
+		assertEquals( 1, sessionFactory().getStatistics().getPrepareStatementCount() );
+		sessionFactory().getStatistics().clear();
+
+		// all collections should be initialized now
+		for (EmployeeGroup group : groups) {
```

Comment:
Before the fix, the following fails for the second collection.

After the fix, there is no failure because both collections were properly initialized by subselect fetching.